panel/widgets/chat.py

```diff
@@ -0,0 +1,1447 @@
+            The avatar to use; overrides the message entry's avatar if provided.
+        respond : bool
+            Whether to execute the callback.
+
+        Returns
+        -------
+        The entry that was created.
+        """
+        entry = self._build_entry(value, user=user, avatar=avatar)
+        self._chat_log.append(entry)
+
+        if respond:
+            self.respond()
+        return entry
+
+    def stream(
+            self,
+            value: str,
+            user: Optional[str] = None,
+            avatar: Optional[Union[str, BinaryIO]] = None,
+            entry: Optional[ChatEntry] = None,
+        ) -> ChatEntry:
+        """
+        Streams a token and updates the provided entry, if provided.
+        Otherwise creates a new entry in the chat log, so be sure the
+        returned entry is passed back into the method, e.g.
+        `entry = chat.stream(token, entry=entry)`.
+
+        This method is primarily for outputs that are not generators--
+        notably LangChain. For most cases, use the send method instead.
+
+        Parameters
+        ----------
+        value : str
+            The new token value to stream.
+        user : Optional[str]
+            The user to stream as; overrides the entry's user if provided.
+        avatar : Optional[Union[str, BinaryIO]]
+            The avatar to use; overrides the entry's avatar if provided.
+        entry : Optional[ChatEntry]
+            The entry to update.
+
+        Returns
+        -------
+        The entry that was updated.
+        """
+        replace = entry is None
+        entry = self._build_entry(entry or value, user=user, avatar=avatar)
+        if replace:
+            self._replace_placeholder(entry)
+        else:
+            self._stream(value, entry)
+        return entry
+
+    def respond(self):
+        """
+        Executes the callback with the latest entry in the chat log.
+        """
+        self._callback_trigger.param.trigger("clicks")
+
+    def undo(self, count: Optional[int] = 1) -> List[Any]:
+        """
+        Removes the last `count` of entries from
+        the chat log and returns them.
+
+        Parameters
+        ----------
+        count : Optional[int]
+            The number of entries to remove, starting from the last entry.
+
+        Returns
+        -------
+        The entries that were removed.
+        """
+        if count == 0:
+            return []
+        entries = self._chat_log.objects
+        undone_entries = entries[-count:]
+        self._chat_log.objects = entries[:-count]
+        return undone_entries
+
+    def clear(self) -> List[Any]:
+        """
+        Clears the chat log and returns
+        the entries that were cleared.
+
+        Returns
+        -------
+        The entries that were cleared.
+        """
+        cleared_entries = self._chat_log.objects
+        self._chat_log.clear()
+        return cleared_entries
+
+
+class ChatInterface(ChatFeed):
```

Comment:
Should I expect to be able use custom widgets made by `Viewer` or `ReactiveHTML` with the `ChatInterface`?

An example is the below that surfaces some issues

- If you click send an error is raised.
- The `ChatInterface` is not layed out nicely. Here I would expect to be able to position the action buttons above my widget(s). I.e. in a `Column` instead of a `Row`. And I would expect the whole app to be within the visible window on start up.

My experience is that its really hard in Panel to create custom widgets using `Viewer` or `ReactiveHTML`. A widget should in most cases be any `Parameterized` with a .value parameter. But most often its something inheriting from `Widget`. How can you make fully accepted custom widgets in Panel using `Viewer` or `ReactiveHTML` @philippjfr ?

![image](https://github.com/holoviz/panel/assets/42288570/5e09e4c7-da9d-43ab-b8e2-7fd9969aaad8)

and below

![image](https://github.com/holoviz/panel/assets/42288570/7a34ab08-c195-43d0-9a80-1aae56a0bc16)


The error when clicking send is 


```bash
  File "/home/jovyan/repos/private/panel/panel/widgets/chat.py", line 1304, in _click_send
    value = active_widget.value
AttributeError: 'Column' object has no attribute 'value'
```

The code is

```python
import panel as pn
from panel.widgets import ChatEntry, ChatInterface
import param

pn.extension("codeeditor")

source_framework = "Streamlit"
target_framework = "Panel"

ASSISTANT_AVATAR = "🤖"
USER_AVATAR = "😃"

EXAMPLES = {
    "hello_world": {
        "streamlit": """
import streamlit as st

st.write("Hello World")
"""
    }
}

class InputWidget(pn.viewable.Viewer):
    value = param.String("Please convert the Streamlit code to Panel code")
    source = param.String(EXAMPLES["hello_world"]["streamlit"])
    target = param.String()

    def __init__(self, **params):
        super().__init__(**params)

        org_sizing_mode = pn.config.sizing_mode
        pn.config.sizing_mode = "stretch_width"

        self._text_input = pn.widgets.TextAreaInput.from_param(self.param.value, name="Task", height=200)
        self._source_code = pn.widgets.CodeEditor.from_param(self.param.source, name="Source Code", height=400, theme="monokai")
        self._target_code = pn.widgets.CodeEditor.from_param(self.param.target, name="Target Code", height=400, theme="monokai")
        self._panel = pn.Column(
            self._text_input, pn.Row(pn.Column(f"{source_framework} code", self._source_code), pn.Column(f"{target_framework} code", self._target_code)),
        )

        pn.config.sizing_mode = org_sizing_mode
        
    
    def __panel__(self):
        return self._panel

    async def callback(self, contents, user, instance):
        yield contents

def assistant(value)->ChatEntry:
    return ChatEntry(value=value, user="Assistant", avatar=ASSISTANT_AVATAR)

def user(value)->ChatEntry:
    return ChatEntry(value=value, user="User", avatar=USER_AVATAR)

input_widget = InputWidget()

chat = ChatInterface(value=[assistant(f"Hi. I am a {source_framework} to {target_framework} data app converter assistant.")], widgets=[input_widget], callback=input_widget.callback)

chat.servable()
```

Reply:
Another small issue is that there is no predefined and configurable dictionary of usernames and avatars such that a ChatEntry knows how to automatically configure the `Assistant` avatar if the name is `Assistant` and `System` avatar if the name is `System` etc. It just means I have to spend time on looking up useful avatars and adding them to my code. Its friction.

Could the `ChatEntry` have an `avatars` `Dict` parameter that contains combinations of user name and avatar for the most common user names? And then let it be customizable via

```
ChatEntry.avatars = {"assistant": "👧", ...}
# and
ChatEntry.avatars["bot"] = "🤷‍♀️"
```

The look up of avatar should be based on lower cased user names.

Reply:
Ok. Maybe for my example the answer is to provide the `TextAreaInput` to the `ChatInterface` and lay that out in a Column with the Editors in a Row below?

But it still needs to be able to align nicely inside the visible window on start up.

Reply:
> Should I expect to be able use custom widgets made by Viewer or ReactiveHTML with the ChatInterface?

I believe that is what ChatFeed is for if they require that level of customizability.

I also don't understand what sending the Streamlit code on the left and Panel code on the right should do?

In any case I believe this PR is large enough and I don't want to hold it up any longer with these potential use cases. Perhaps you can submit an issue for further improvements down the road?

Reply:
Added default user_avatars
```python
import panel as pn

pn.extension()


def callback(contents, user, instance):
    yield {"user": "ChatGPT", "value": f"This isn't really ChatGPT; I only echo: {contents}"}


pn.template.FastListTemplate(
    main=[
        pn.widgets.ChatInterface(
            callback=callback,
        )
    ]
).servable()

```

<img width="1298" alt="image" src="https://github.com/holoviz/panel/assets/15331990/a73c3851-9e47-47cf-942a-0e8e9a1b374f">